pkg/orchestrator/scheduler/batch_service_job.go

```diff
@@ -89,19 +91,42 @@ func (b *BatchServiceJobScheduler) Process(ctx context.Context, evaluation *mode
 	execsByApprovalStatus.toApprove.markApproved(plan)
 	execsByApprovalStatus.toReject.markStopped(execRejected, plan)
 
-	// create new executions if needed
-	remainingExecutionCount := desiredRemainingCount - execsByApprovalStatus.activeCount()
-	if remainingExecutionCount > 0 {
-		allFailed := existingExecs.filterFailed().union(lost)
-		var placementErr error
-		if len(allFailed) > 0 && !b.retryStrategy.ShouldRetry(ctx, orchestrator.RetryRequest{JobID: job.ID}) {
-			placementErr = fmt.Errorf("exceeded max retries for job %s", job.ID)
-		} else {
-			_, placementErr = b.createMissingExecs(ctx, remainingExecutionCount, &job, plan)
+	// How many executions failed due to compute nodes rejecting bids?
+	rejectedExecutions := 0
+
+	for _, execution := range jobExecutions {
+		if execution.ComputeState.StateType == models.ExecutionStateAskForBidRejected {
+			// This execution was rejected by its compute node
+			rejectedExecutions = rejectedExecutions + 1
 		}
-		if placementErr != nil {
-			b.handleFailure(nonTerminalExecs, allFailed, plan, placementErr)
-			return b.planner.Process(ctx, plan)
+	}
+
+	if (rejectedExecutions > 0) && evaluation.TriggeredBy != models.EvalTriggerDefer {
+		// If we had failed executions due to bid rejections in the
+		// past, then we should retry. This causes the scheduler to be
+		// nivoked again after the retry delay; when that happens,
+		// evaluation.TriggeredBy will equal models.EvalTriggerDefer so
+		// the test above checks for that to make us not retry *again*
+		// after we've retried.
+		b.handleRetry(plan, &job, rejectedExecutions)
+		return b.planner.Process(ctx, plan)
+	} else {
```

Comment:
I am not sure I follow this. An evaluation arriving at the scheduler should be processed as is. If an delayed evaluation was enqueued, then the evaluation would arrive at the scheduler after the backoff duration.

Reply:
Yes. I'll improve the comment - this is because a retry involves two evaluations: one that diagnoses the situation as a failure that should be retried (which calls `b.handleRetry` to schedule that retry evaluation). When the retry evaluation is then processed, after the delay, it finds itself in exactly the same situation, because nothing else has changed - but it can observe that it is a retry evaluation (`evaluation.TriggeredBy == models.EvalTriggerDefer`) so it should actually plan new execution attempts rather than deferring itself once more.

Reply:
As explained in our last meeting, the evaluation broker is deduping evaluations for the same job, and there is no guarantee that the deferred evaluation is the one the will be served. Relying on the `TriggeredBy` field to make scheduling decisions might result in unexpected behaviour.

**Second,** `rejected` executions is not the only reason that we want to retry scheduling. Failed executions, or no live nodes at all are reasons where we want to queue the job and retry at a later time

**Third,** we shouldn't delay a job if there are `rejected` or `failed` executions, but we should delay if the scheduler couldn't find any new node at allocate the execution to, and the retry policy still says we can retry. If I am following the code correctly, this is what your scheduler will do:
1. found 10 nodes, ask 1 node to bid
2. node rejects
3. backoff and retry later instead of asking the remaining 9 nodes immediately

Though the expected behaviour is:
1. found 10 nodes, ask 1 node to bid
2. node rejects
3. found 9 nodes, ask 1 node to bid
4. ...
5. found 0 nodes, backoff and retry later

We used to ask more than one node in each iteration, but that is no [longer working](https://github.com/bacalhau-project/expanso-planning/issues/597)

This means the node ranking should exclude nodes that were asked before, and somehow reconsider them later when retrying after backoff.

Reply:
> As explained in our last meeting, the evaluation broker is deduping evaluations for the same job, and there is no guarantee that the deferred evaluation is the one the will be served. Relying on the `TriggeredBy` field to make scheduling decisions might result in unexpected behaviour.

I think it might end up having evaluations that do nothing, and might end up waiting longer than it needs to if a `WaitUntil` evaluation is lost because an immediate one overwrites it in dedupe (what exactly is the logic by which the evaluation broker decides which one to keep when deduping?), but I think it should always make progress: any new evaluation will create a delayed evaluation that will then make progress.

The question is, what could we do better?

> **Second,** `rejected` executions is not the only reason that we want to retry scheduling. Failed executions, or no live nodes at all are reasons where we want to queue the job and retry at a later time

Yeah. To refactor the retry strategy, I had to combine the handling of rejected, failed, and lost executions so I think this is now OK. But I only made it delay if there were rejected executions, because we didn't originally have a delay for the other cases.

> **Third,** we shouldn't delay a job if there are `rejected` or `failed` executions, but we should delay if the scheduler couldn't find any new node at allocate the execution to, and the retry policy still says we can retry. If I am following the code correctly, this is what your scheduler will do:

Originally the code worked more like that - `createMissingExecs` used to return an error if it couldn't find a suitable node - but this stopped working when I rebased onto `main` after Christmas :-(

So. What can we do better? The tricky thing is that the scheduler has little information to go on: the desired state of the job and the execution history. When it's triggered because an execution was rejected by a node, if it decides it needs to wait before trying, it creates a new execution with a `WaitUntil`, but when THAT execution occurs, how does it know to do anything different to the original execution?

Perhaps rather than looking at `TriggeredBy`, it could look at the timestamps in the execution history. Perhaps if the most recently rejected execution is less than the current retry delay ago we should just create a delayed evaluation (delayed just enough to wait until the current retry delay later than the rejection time of the last execution, so if an evaluation rocks up after 5s for some reason and we only need to try again in 8s, we create one delayed by 3s). If the evaluation broker dedupes by always keeping the latest evaluation for a given job ID, this might work better?

What do you think?

> We used to ask more than one node in each iteration, but that is no [longer working](https://github.com/bacalhau-project/expanso-planning/issues/597)

Hmmm, I thought we created as many executions as we wanted in one go, from looking at how `createMissingExecs` is used and works!

> This means the node ranking should exclude nodes that were asked before, and somehow reconsider them later when retrying after backoff.

Ok, I made the node ranking prefer nodes that haven't been asked before, and accept (at a lower priority) nodes that have been asked before, under the assumption that this meant the fresh nodes would be tried in preference but if we rant out of them we'd start looking again at previously tried nodes.

Should we try to NOT introduce delays if there are "fresh" nodes still to try, only delaying when we have to go back to previously-tried nodes?

That would require deeper refactoring of the batch/service job scheduler, i think. I'm nervous about going into deeper depths of yak-shaving as this thing has dragged on a lot already. Would it be good to get something basic and working merged into main, by me writing tests and documentation for what I have, then make it more clever and less wasteful in new branches?